c/src/lib/libbsp/arm/shared/stm32fxxxx/i2c/hal-i2c.c

```diff
@@ -0,0 +1,331 @@
+					break;
+	case I2C_THREE:	return (IRQn_Type)GET_I2C_VEC(I2C_THREE++);
+					break;
+	case I2C_FOUR:	return (IRQn_Type)GET_I2C_VEC(I2C_FOUR++);
+					break;
+	}
+}
+
+/**
+ * Initialize I2C clock
+ */
+void stm32_i2c_initialize_i2c_clock(I2C_Instance i2c_instance)
+{
+	switch(i2c_instance)
+	{
+	case I2C_ONE:	__HAL_RCC_I2C1_CLK_ENABLE();
+					break;
+	case I2C_TWO:	__HAL_RCC_I2C2_CLK_ENABLE();
+					break;
+	case I2C_THREE:	__HAL_RCC_I2C3_CLK_ENABLE();
+					break;
+#if (STM32_ENABLE_I2C4)
+	case I2C_FOUR:	__HAL_RCC_I2C4_CLK_ENABLE();
+					break;
+#endif
+	}
+}
+
+/**
+ * Returns the base address of the I2C instances
+ */
+I2C_TypeDef* stm32_i2c_get_i2c_instance(I2C_Instance i2c_instance)
+{
+	switch(i2c_instance)
+	{
+	case I2C_ONE:	return I2C1;
+					break;
+	case I2C_TWO:	return I2C2;
+					break;
+	case I2C_THREE:	return I2C3;
+					break;
+#if (STM32_ENABLE_I2C4)
+	case I2C_FOUR:	return I2C4;
+					break;
+#endif
+	}
+}
+
+/**
+ * I2C Interrupt Event ISR
+ */
+static void stm32_i2c_event_irq(void* arg)
+{
+	rtems_status_code sc;
+	stm32_i2c_bus *bus;
+
+	HAL_I2C_EV_IRQHandler((I2C_HandleTypeDef *) arg);
+
+	sc = rtems_event_transient_send(bus->task_id);
+	_Assert(sc == RTEMS_SUCCESSFUL);
+}
+
+/**
+ * I2C Initiate Transfer
+ */
+static int stm32_i2c_transfer(
+  i2c_bus *base,
+  i2c_msg *msgs,
+  uint32_t msg_count
+)
+{
+	rtems_status_code sc = RTEMS_SUCCESSFUL;
+	stm32_i2c_bus *bus = (stm32_i2c_bus *) base;
+
+	i2c_msg *message = msgs->buf;
+
+	/* Wait for the I2C bus to be ready */
+	while (HAL_I2C_GetState(&bus->handle) != HAL_I2C_STATE_READY);
+
+	if(msgs->flags == 0)
+	{
+		if(HAL_I2C_Master_Transmit_IT(&bus->handle, (uint16_t)message->addr, (uint8_t*)message->buf, msgs->len)!= HAL_OK){
+			return -1;
+		}
+	}
+	else if(msgs->flags == I2C_M_RD)
+	{
+		if(HAL_I2C_Master_Receive_IT(&bus->handle, (uint16_t)message->addr, (uint8_t*)message->buf, msgs->len) != HAL_OK){
+			return -1;
+		}
+	}
+
+	bus->task_id = rtems_task_self();
+
+	sc = rtems_event_transient_receive(RTEMS_WAIT, 20);
+
+	if (sc != RTEMS_SUCCESSFUL) {
+		return -ETIMEDOUT;
+	}
+}
+
+/**
+ * I2C Bus Destroy and De-initialise
+  */
+static int stm32_i2c_deinit_destroy(stm32_i2c_bus * bus)
+{
+	rtems_status_code sc;
+
+	sc = rtems_interrupt_handler_remove(stm32_get_i2c_vec_num(bus->instance), stm32_i2c_event_irq, &bus->handle);
+	_Assert(sc == RTEMS_SUCCESSFUL);
+	(void) sc;
+
+	i2c_bus_destroy_and_free(&bus->base);
+	return (HAL_I2C_DeInit(&bus->handle));
+}
+
+/**
+ * Initialize GPIO pins
+ */
+void stm32_i2c_gpio_init(stm32_i2c_bus * bus)
+{
+
+	GPIO_InitTypeDef  GPIO_InitStruct;
+
+	stm32_i2c_initialize_i2c_clock(bus->instance);
+
+	/* Initialize GPIO Clock */
+	__HAL_RCC_GPIOA_CLK_ENABLE();
+	__HAL_RCC_GPIOB_CLK_ENABLE();
+	__HAL_RCC_GPIOC_CLK_ENABLE();
+
+	GPIO_InitStruct.Mode      = GPIO_MODE_AF_OD;
+	GPIO_InitStruct.Pull      = GPIO_PULLUP;
+	GPIO_InitStruct.Speed     = GPIO_SPEED_FAST;
+	GPIO_InitStruct.Alternate = i2c_pin_config[bus->instance].alternate_func;
```

Comment:
Probably future work, but make the peripheral libraries use the GPIO library to try to avoid pin conflicts by using the GPIO group functionality.